File: ics_smb_management/models/smb_dashboard.py
# -*- coding: utf-8 -*-
# Part of Odoo. See LICENSE file for full copyright and licensing details.

from odoo import api, fields, models
from datetime import datetime, timedelta
from dateutil.relativedelta import relativedelta


class SmbDashboard(models.Model):
    _name = 'smb.dashboard'
    _description = 'SMB Dashboard'

    name = fields.Char(string='Dashboard', default='SMB Dashboard')

    @api.model
    def get_smb_statistics(self):
        """Statistics for SMB dashboard: credit, orders, collection, projects."""
        SaleOrder = self.env['sale.order']
        today = fields.Date.context_today(self)

        # Credit workflow
        pending_credit = SaleOrder.search_count([
            ('smb_credit_state', '=', 'sent_to_credit'),
            ('state', 'in', ('draft', 'sent')),
        ])
        credit_approved = SaleOrder.search_count([
            ('smb_credit_state', '=', 'credit_approved'),
        ])
        credit_rejected = SaleOrder.search_count([
            ('smb_credit_state', '=', 'credit_rejected'),
        ])
        total_quotations = SaleOrder.search_count([
            ('state', 'in', ('draft', 'sent')),
        ])
        confirmed_orders = SaleOrder.search_count([
            ('state', 'in', ('sale', 'done')),
        ])

        # Overdue invoices (collection)
        overdue_invoices = 0
        if 'account.move' in self.env:
            overdue_invoices = self.env['account.move'].search_count([
                ('move_type', '=', 'out_invoice'),
                ('state', '=', 'posted'),
                ('payment_state', 'in', ('not_paid', 'partial')),
                ('invoice_date_due', '<', today),
            ])

        # Delivery projects (projects linked to SO)
        delivery_projects = 0
        if 'project.project' in self.env:
            delivery_projects = self.env['project.project'].search_count([
                ('sale_order_id', '!=', False),
            ])

        # Credit state distribution (for chart)
        credit_distribution = self._get_credit_state_distribution(SaleOrder)
        # Monthly trend: orders confirmed last 6 months
        monthly_trend = self._get_monthly_trend(SaleOrder)

        return {
            'pending_credit': pending_credit,
            'credit_approved': credit_approved,
            'credit_rejected': credit_rejected,
            'total_quotations': total_quotations,
            'confirmed_orders': confirmed_orders,
            'overdue_invoices': overdue_invoices,
            'delivery_projects': delivery_projects,
            'credit_distribution': credit_distribution,
            'monthly_trend': monthly_trend,
        }

    def _get_credit_state_distribution(self, sale_order_obj):
        labels = []
        values = []
        for state, label in [
            ('sent_to_credit', 'Pending Credit'),
            ('credit_approved', 'Credit Approved'),
            ('credit_rejected', 'Credit Rejected'),
        ]:
            count = sale_order_obj.search_count([('smb_credit_state', '=', state)])
            if count > 0:
                labels.append(label)
                values.append(count)
        return {'labels': labels, 'values': values}

    def _get_monthly_trend(self, sale_order_obj):
        months = []
        values = []
        today = datetime.today()
        for i in range(5, -1, -1):
            month_start = (today - relativedelta(months=i)).replace(day=1)
            month_end = (month_start + relativedelta(months=1)) - timedelta(days=1)
            count = sale_order_obj.search_count([
                ('state', 'in', ('sale', 'done')),
                ('date_order', '>=', month_start.strftime('%Y-%m-%d')),
                ('date_order', '<=', month_end.strftime('%Y-%m-%d')),
            ])
            months.append(month_start.strftime('%b'))
            values.append(count)
        return {'labels': months, 'values': values}

File: ics_smb_management/static/src/js/smb_dashboard.js
/** @odoo-module **/

import { registry } from "@web/core/registry";
import { Component, useState, onWillStart, onMounted, useRef } from "@odoo/owl";
import { useService } from "@web/core/utils/hooks";
import { loadJS } from "@web/core/assets";

export class SmbDashboard extends Component {
    setup() {
        this.orm = useService("orm");
        this.action = useService("action");

        this.state = useState({
            stats: {
                pending_credit: 0,
                credit_approved: 0,
                credit_rejected: 0,
                total_quotations: 0,
                confirmed_orders: 0,
                overdue_invoices: 0,
                delivery_projects: 0,
                credit_distribution: { labels: [], values: [] },
                monthly_trend: { labels: [], values: [] },
            },
            isLoading: true,
        });

        this.creditChart = useRef("creditChart");
        this.monthlyTrendChart = useRef("monthlyTrendChart");

        onWillStart(async () => {
            await loadJS("/web/static/lib/Chart/Chart.js");
            await this.loadDashboardData();
        });

        onMounted(() => {
            this.renderCharts();
        });
    }

    async loadDashboardData() {
        try {
            const data = await this.orm.call(
                "smb.dashboard",
                "get_smb_statistics",
                []
            );
            if (data) {
                this.state.stats = data;
            }
        } catch (error) {
            console.error("Error loading SMB dashboard data:", error);
        } finally {
            this.state.isLoading = false;
        }
    }

    renderCharts() {
        this.renderCreditChart();
        this.renderMonthlyTrendChart();
    }

    renderCreditChart() {
        if (
            this.creditChart.el &&
            this.state.stats.credit_distribution &&
            this.state.stats.credit_distribution.values.length > 0
        ) {
            const ctx = this.creditChart.el.getContext("2d");
            new Chart(ctx, {
                type: "doughnut",
                data: {
                    labels: this.state.stats.credit_distribution.labels,
                    datasets: [
                        {
                            data: this.state.stats.credit_distribution.values,
                            backgroundColor: ["#f6c23e", "#1cc88a", "#e74a3b"],
                            hoverBackgroundColor: ["#dda20a", "#17a673", "#be2617"],
                            borderWidth: 2,
                        },
                    ],
                },
                options: {
                    responsive: true,
                    maintainAspectRatio: false,
                    plugins: {
                        legend: { display: true, position: "bottom" },
                    },
                },
            });
        }
    }

    renderMonthlyTrendChart() {
        if (
            this.monthlyTrendChart.el &&
            this.state.stats.monthly_trend &&
            this.state.stats.monthly_trend.labels.length > 0
        ) {
            const ctx = this.monthlyTrendChart.el.getContext("2d");
            new Chart(ctx, {
                type: "line",
                data: {
                    labels: this.state.stats.monthly_trend.labels,
                    datasets: [
                        {
                            label: "Confirmed Orders",
                            data: this.state.stats.monthly_trend.values,
                            borderColor: "#4e73df",
                            backgroundColor: "rgba(78, 115, 223, 0.1)",
                            fill: true,
                            tension: 0.3,
                        },
                    ],
                },
                options: {
                    responsive: true,
                    maintainAspectRatio: false,
                    scales: {
                        y: {
                            beginAtZero: true,
                            ticks: { precision: 0 },
                        },
                    },
                    plugins: {
                        legend: { display: false },
                    },
                },
            });
        }
    }

    openPendingCredit() {
        this.action.doAction({
            name: "Credit Approval Queue",
            type: "ir.actions.act_window",
            res_model: "sale.order",
            views: [
                [false, "list"],
                [false, "form"],
            ],
            domain: [["smb_credit_state", "=", "sent_to_credit"]],
            context: { search_default_sent_to_credit: 1 },
        });
    }

    openCreditApproved() {
        this.action.doAction({
            name: "Credit Approved",
            type: "ir.actions.act_window",
            res_model: "sale.order",
            views: [
                [false, "list"],
                [false, "form"],
            ],
            domain: [["smb_credit_state", "=", "credit_approved"]],
        });
    }

    openCreditRejected() {
        this.action.doAction({
            name: "Credit Rejected",
            type: "ir.actions.act_window",
            res_model: "sale.order",
            views: [
                [false, "list"],
                [false, "form"],
            ],
            domain: [["smb_credit_state", "=", "credit_rejected"]],
        });
    }

    openOverdueInvoices() {
        this.action.doAction({
            name: "Overdue Invoices",
            type: "ir.actions.act_window",
            res_model: "account.move",
            views: [
                [false, "list"],
                [false, "form"],
            ],
            domain: [
                ["move_type", "=", "out_invoice"],
                ["state", "=", "posted"],
                ["payment_state", "in", ["not_paid", "partial"]],
            ],
        });
    }

    openQuotations() {
        this.action.doAction({
            name: "Quotations",
            type: "ir.actions.act_window",
            res_model: "sale.order",
            views: [
                [false, "list"],
                [false, "form"],
            ],
            domain: [["state", "in", ["draft", "sent"]]],
        });
    }

    openConfirmedOrders() {
        this.action.doAction({
            name: "Confirmed Orders",
            type: "ir.actions.act_window",
            res_model: "sale.order",
            views: [
                [false, "list"],
                [false, "form"],
            ],
            domain: [["state", "in", ["sale", "done"]]],
        });
    }

    openDeliveryProjects() {
        this.action.doAction({
            name: "Delivery Projects",
            type: "ir.actions.act_window",
            res_model: "project.project",
            views: [
                [false, "list"],
                [false, "form"],
            ],
            domain: [["sale_order_id", "!=", false]],
        });
    }
}

SmbDashboard.template = "ics_smb_management.SmbDashboard";

registry.category("actions").add("ics_smb_management.smb_dashboard", SmbDashboard);
